benchmarks/distributed/ddp/compare/python_ddp.py

```diff
@@ -0,0 +1,228 @@
+            assert bucket.buffer is not None, 'bucket.buffer should not be None'
+        print("len(param_to_bucket)={}, len(buckets)={}".format(
+            len(params_to_buckets), len(buckets)))
+
+        # Sanity check to ensure all params are distributed correctly into buckets
+        total_params_in_buckets = 0
+        for bucket in buckets:
+            total_params_in_buckets += len(bucket.param_to_offset)
+        assert total_param == total_params_in_buckets
+
+        return params_to_buckets, buckets
+
+    # Callback when param.grad is ready. Note during callback, param.grad won't
+    # be ready yet, we MUST use the given ''grad'' which would be passed upon
+    # callback.
+    def _on_param_grad_ready(self, param, grad):
+        """
+        Callback when grad for param is ready. Copy grad to its corresponding
+        bucket. When the bucket is full, kickoff an async all_reduce if
+        async_reduction is set, and adds the resultant handle to
+        self.async_handles.
+
+        .. warning::
+            Note param.grad isn't set yet. Use the passed grad instead.
+        """
+        # Validate bucket and offset are set.
+        bucket = self.param_to_bucket.get(param)
+        assert bucket is not None, "Failed to find bucket for param"
+        offset = bucket.param_to_offset.get(param)
+        assert offset is not None, "offset must be set for param"
+        assert bucket.buffer is not None, "buffer must be allocated"
+
+        # Copy grad to bucket, note param.grad isn't ready yet.
+        sz = param.numel()
+        assert grad is not None
+        assert param.requires_grad
+        assert param.numel() == grad.numel()
+        bucket.buffer[offset : offset + sz].copy_(grad.data.view(-1))
+        bucket.ready_param_grad_count += 1
+
+        # Kickoff grad reduction async when bucket is full. This ensures grad
+        # reduction and other grad calculation runs in parallel.
+        if self.async_reduction and bucket.is_full():
+            bucket.buffer.div_(self.world_size)
+            handle = dist.all_reduce(
+                bucket.buffer, dist.ReduceOp.SUM, self.process_group, True)
+            self.async_handles.add(handle)
+
+    def forward(self, *inputs, **kwargs):
+        return self.module(*inputs, **kwargs)
+
+    def all_reduce_grads(self):
+        """
+        Reduces all gradients across worker and updates param gradients. The
+        client should call this func post backward.
+
+        If async_reduction is True, waits for all async handles (of all_reduce),
+        otherwise, kicks off synchrous all_reduce for all buckets.
+
+        Once all all buckets are reduced, copy the reduced grads back to their
+        original parameters. After that, reset all buckets in prep for the next
+        iteration.
+        """
+        if self.async_reduction:
+            for handle in self.async_handles:
+                handle.wait()
+            self.async_handles.clear()
```

Comment:
If `async_handles` is a list of a fixed length, instead of a set, here we can just use `torch.futures.wait_all(async_handles)`, although I don't think it will make a difference in performance.